chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,949 @@
+                        }))
+                    }
+                    Some(ReshardingEventType::Split(split)) => {
+                        // It's fine only if this shard is already a child of the existing event.
+                        if split.left_child_shard.shard_id() != shard_id
+                            && split.right_child_shard.shard_id() != shard_id
+                        {
+                            return error_two_reshardings();
+                        }
+                    }
+                }
+            }
+        }
+    }
+    // We must have found at least one resharding event by now.
+    event.ok_or_else(|| {
+        let err_msg = "no supported shard layout change found!";
+        error!(target: "resharding", ?shard_layout, err_msg);
+        Error::ReshardingError(err_msg.to_owned())
+    })
+}
+
+/// Task to perform the actual split of a flat storage shard. This may be a long operation time-wise.
+///
+/// Conceptually it simply copies each key-value pair from the parent shard to the correct child.
+fn split_shard_task(
+    resharder: FlatStorageResharderInner,
+    controller: FlatStorageResharderController,
+) {
+    let success = split_shard_task_impl(resharder.clone(), controller.clone());
+    split_shard_task_postprocessing(resharder, success);
+    info!(target: "resharding", "flat storage shard split task finished, success: {success}");
+    if let Err(err) = controller.completion_sender.send(success) {
+        warn!(target: "resharding", ?err, "error notifying completion of flat storage shard split task")
+    };
+}
+
+/// Retrieve parent shard UIds and current resharding event status.
+/// Resharding event must be of type "Split".
+fn get_parent_shard_and_status(
+    resharder: &FlatStorageResharderInner,
+) -> (ShardUId, SplittingParentStatus) {
+    let event = resharder.resharding_event.lock().unwrap();
+    match event.as_ref() {
+        Some(FlatStorageReshardingEvent::Split(parent_shard, status)) => {
+            (*parent_shard, status.clone())
+        }
+        None => panic!("a resharding event must exist!"),
+    }
+}
+
+/// Performs the bulk of [split_shard_task].
+///
+/// Returns `true` if the routine completed successfully.
+fn split_shard_task_impl(
+    resharder: FlatStorageResharderInner,
+    controller: FlatStorageResharderController,
+) -> bool {
+    if controller.is_interrupted() {
+        return false;
+    }
+
+    /// Determines after how many key-values the process stops to
+    /// commit changes and to check interruptions.
+    const BATCH_SIZE: usize = 10_000;
+
+    let (parent_shard, status) = get_parent_shard_and_status(&resharder);
+
+    info!(target: "resharding", ?parent_shard, "flat storage shard split task: starting key-values copy");
+
+    // Prepare the store object for commits and the iterator over parent's flat storage.
+    let flat_store = resharder.runtime.store().flat_store();
+    let mut iter = flat_store.iter(parent_shard);
```

Comment:
IIUC the flat storage of the parent is "locked" as far as state sync is concerned by the change of state.
There's in fact a requirement that the flat storage height should be equal to the chain head; then there wouldn't be any further deltas to apply because the nothing will write to the parent shard after this point.

Reply:
I think what Wac means here is that we need to check the following
- Flat head is at height which is the first block of the new epoch with the new shard layout
- Flat head is indeed locked/frozen

Btw, I don't think the assumption where we say flat head is same as chain head is true. Flat head when locked can be behind chain head.

Reply:
I see! I added a TODO to check / bring flat storage to the right height. I'll need to write a test too.